check_sophos_xg_info.pl

```diff
@@ -60,20 +60,81 @@ BEGIN
 );
 
 $mp->add_arg(
-    spec => 'hostname|H=s',
-    help => '',
+    spec     => 'hostname|H=s',
+    help     => 'Hostname or IP of the device with SNMP access enabled',
     required => 1
 );
 
-$mp->getopts;
+$mp->add_arg(
+    spec => 'username|u=s',
+    help => 'Username for SNMPv3',
+);
+
+$mp->add_arg(
+    spec => 'authpassword|A=s',
+    help => 'Authentication protocol password',
+);
+
+$mp->add_arg(
+    spec    => 'authprotocol|a=s',
+    help    => 'Authentication protocol: MD5, SHA, SHA-224, SHA-256, SHA-384, SHA-512 (Default: MD5)',
+    default => 'md5',
+);
 
-#Open SNMP Session
-my ($session, $error) = Net::SNMP->session(
-    -hostname => $mp->opts->hostname,
-    -version => 'snmpv2c',
-    -community => $mp->opts->community,
+$mp->add_arg(
+    spec => 'privpassword|X=s',
+    help => 'privacy protocol password',
 );
 
+$mp->add_arg(
+    spec    => 'privprotocol|x=s',
+    help    => 'Privacy protocol: DES, AES (Default: DES)',
+    default => 'des',
+);
+
+$mp->add_arg(
+    spec    => 'verbose',
+    help    => 'Print verbose/debug information',
+    default => 0,
+);
+
+$mp->getopts;
+
+
+my ($session, $error);
+if (defined($mp->opts->username) && defined($mp->opts->authpassword)) {
+    # SNMPv3 login
+    verb('SNMPv3 login');
+    if (!defined($mp->opts->privpassword)) {
+        verb('SNMPv3 AuthNoPriv login : %s, %s', $mp->opts->username, $mp->opts->authprotocol);
+        ($session, $error) = Net::SNMP->session(
+            -hostname => $mp->opts->hostname,
+            -version => 'snmpv3',
+            -username => $mp->opts->username,
+            -authprotocol => $mp->opts->authprotocol,
+            -authpassword => $mp->opts->authpassword,
+        );
+    } else {
+        verb('SNMPv3 AuthPriv login : %s, %s, %s', ${mp->opts->username}, ${mp->opts->authprotocol}, ${mp->opts->privprotocol});
```

Comment:
You have to remove the curly braces, it should be `$mp->opts->username` instead of `${mp->opts->username}`. And the same for the remaining arguments in the subroutine.